lib/autoload/course/conditional/conditional_dependency_graph.rb

```diff
@@ -0,0 +1,70 @@
+# Dependency graph to handle and resolve the dependencies for the Conditional-Condition framework.
+class Course::Conditional::ConditionalDependencyGraph
+  class EdgeSet < Hash
+    def initialize
+      super { |h, k| h[k] = Set.new }
+    end
+
+    def add(source, edge)
+      self[source] << edge
+    end
+
+    def satisfied_edges_for(source, course_user)
+      self[source].select { |c| c.satisfied_by?(course_user) }
+    end
+  end
+
+  # Initialize a topological sorted conditional dependency graph
+  #
+  # @param [Array<Object] conditionals Array of objects with acts_as_conditional
+  # @raise [ArgumentError] When there is a cyclic dependency within the given conditionals
+  def initialize(conditionals)
+    @edges = EdgeSet.new
+
+    begin
+      @dependency_graph = TSort.tsort(each_node(conditionals), each_child)
+    rescue TSort::Cyclic
+      raise ArgumentError, 'Cyclic dependency detected in given conditionals'
+    end
+  end
+
+  # Walk through the dependency graph and resolve all conditionals base on the given satisfied
+  # conditions.
+  #
+  # @param[Array<Condition] satisfied_conditions Conditions that have already been satisfied
```

Comment:
I think there are supposed to be a space before [, not sure if Yardoc can parse this.